File: study-nurio-mobile/ios/Routing/OAuthRouteDecisionHandler.swift
import Foundation
import HotwireNative

@MainActor
final class OAuthRouteDecisionHandler: RouteDecisionHandler {
    let name = "oauth"

    func matches(location: URL, configuration: Navigator.Configuration) -> Bool {
        SocialAuthRoute.resolve(
            startPath: location.absoluteString,
            baseURL: configuration.startLocation
        ) != nil
    }

    func handle(location: URL, configuration: Navigator.Configuration, navigator: Navigator) -> Router.Decision {
        guard let route = SocialAuthRoute.resolve(
            startPath: location.absoluteString,
            baseURL: configuration.startLocation
        ) else {
            return .cancel
        }

        OAuthSessionCoordinator.shared.presentationAnchorProvider = { [weak navigator] in
            navigator?.activeNavigationController.view.window
        }

        SocialAuthCoordinator.shared.start(route: route) { [weak navigator] result in
            SocialAuthResultHandler.handle(
                result,
                presenting: navigator?.activeNavigationController
            )
        }

        return .cancel
    }
}

File: study-nurio-mobile/android/app/src/main/java/com/nurio/study/android/routing/OAuthRouteDecisionHandler.kt
package com.nurio.study.android.routing

import com.nurio.study.android.BuildConfig
import com.nurio.study.android.MainActivity
import com.nurio.study.android.auth.SocialAuthRoute
import dev.hotwire.navigation.activities.HotwireActivity
import dev.hotwire.navigation.navigator.NavigatorConfiguration
import dev.hotwire.navigation.routing.Router

class OAuthRouteDecisionHandler : Router.RouteDecisionHandler {
    override val name = "oauth-browser-tab"

    override fun matches(
        location: String,
        configuration: NavigatorConfiguration
    ): Boolean {
        return SocialAuthRoute.resolve(location, BuildConfig.BASE_URL) != null
    }

    override fun handle(
        location: String,
        configuration: NavigatorConfiguration,
        activity: HotwireActivity
    ): Router.Decision {
        val route = SocialAuthRoute.resolve(location, BuildConfig.BASE_URL)
            ?: return Router.Decision.CANCEL
        (activity as? MainActivity)?.dispatchSocialAuth(route)

        return Router.Decision.CANCEL
    }
}
